internal/devpkg/flakeref.go

```diff
@@ -231,9 +228,111 @@ func parseGitHubFlakeRef(refURL *url.URL, parsed *FlakeRef) error {
 	return nil
 }
 
-// String returns the raw flakeref string as given to ParseFlakeRef.
+// String encodes the flake reference as a URL-like string. It normalizes
+// the result such that if two FlakeRef values are equal, then their strings
+// will also be equal.
+//
+// There are multiple ways to encode a flake reference. String uses the
+// following rules to normalize the result:
+//
+//   - the URL-like form is always used, even for paths and indirects.
+//   - the scheme is always present, even if it's optional.
+//   - paths are cleaned per path.Clean.
+//   - fields that can be put in either the path or the query string are always
+//     put in the path.
+//   - query parameters are sorted by key.
+//
+// If f is missing a type or has any invalid fields, String returns an empty
+// string.
 func (f FlakeRef) String() string {
-	return f.raw
+	switch f.Type {
+	case "file":
+		if f.URL == "" {
+			return ""
+		}
+		return "file+" + f.URL
+	case "git":
+		if f.URL == "" {
+			return ""
+		}
+		if !strings.HasPrefix(f.URL, "git") {
+			f.URL = "git+" + f.URL
+		}
+
+		// Nix removes "ref" and "rev" from the query string
+		// (but not other parameters) after parsing. If they're empty,
+		// we can skip parsing the URL. Otherwise, we need to add them
+		// back.
+		if f.Ref == "" && f.Rev == "" {
+			return f.URL
+		}
+		url, err := url.Parse(f.URL)
+		if err != nil {
+			// This should be rare and only happen if the caller
+			// messed with the parsed URL.
+			return ""
+		}
+		url.RawQuery = buildQueryString("ref", f.Ref, "rev", f.Rev, "dir", f.Dir)
+		return url.String()
+	case "github":
+		if f.Owner == "" || f.Repo == "" {
+			return ""
```

Comment:
should this be an error?

Reply:
Ideally, but I want FlakeRef to be a `fmt.Stringer` so this function is unable to return an error.